### ISSUE TITLE
[BUG] MLFlow Tracking Server doesn’t log experiments when executing the training job in Airflow

### ISSUE THREAD

**Issue by nguyen-dhn** [MLOpsVN/courses.mlops.vn#78]
**Mô tả bug**

- Tên bug: MLFlow Tracking Server doesn’t log experiments when executing the training job in Airflow
- Xảy ra ở bài [[Training Pipeline / Airflow DAG](https://github.com/MLOpsVN/courses.mlops.vn/issues/new?assignees=&labels=bug&template=bug_report.md&title=%5BBUG%5D)]
- Xảy ra ở bước trigger `training_pipeline` in Airflow UI.

**Cách tạo lại bug**

In all repos, make sure all changes have been stashed and your current code base is reflected the following commit hashes:

- Commit hash (mlops-crash-course-code): [bfa5357994a667cabb3cbbf63410eb722f11b5a0](https://github.com/MLOpsVN/mlops-crash-course-code/commit/bfa5357994a667cabb3cbbf63410eb722f11b5a0)
- Commit hash (mlops-crash-course-platform): [25602f81f8a0b978ff8661974909b45489f8015d](https://github.com/MLOpsVN/mlops-crash-course-platform/commit/25602f81f8a0b978ff8661974909b45489f8015d)

---

- In the repo `mlops-crash-course-platform`:
    - Down all Airflow and MLFlow services that are running (remove all volumes)
        
        ```bash
        bash run.sh airflow down --volumes
        bash run.sh mlflow down --volumes
        ```
        
    - Remove all data in `airflow/run_env` and `mlflow/run_dev` (excluded `.gitkeep`)
    - Start Airflow and MLFlow again, make sure all caches have been cleared (no records found in Airflow and MLflow UI)
- Delete all `mlopsvn/mlops_crash_course/training_pipeline` docker images.
- In Airflow UI: login and set the `MLOPS_CRASH_COURSE_CODE_DIR` variable.
- In the repo `mlops-crash-course-code`:
    
    ```bash
    cd training_pipeline
    make build_image
    make deploy_dags
    ```
    
- Trigger DAG (training_pipeline) in Airflow UI. The `model_training_task` are successful but the experiment doesn’t log into MLFlow Tracking Server.

**Attachments**
_(Sometimes you will see MLFlow server's port in my log is 5001. Just ignore this because I cannot run on port 5000 on my machine)._

- The artifact URI in `model_training_task` (Airflow job) is `./artifacts`. Here is the log:
    
    ```bash
    *** Reading local file: /opt/airflow/logs/dag_id=training_pipeline/run_id=scheduled__2022-01-01T00:00:00+00:00/task_id=model_training_task/attempt=1.log
    [2022-11-06, 08:09:27 +07] {taskinstance.py:1171} INFO - Dependencies all met for <TaskInstance: training_pipeline.model_training_task scheduled__2022-01-01T00:00:00+00:00 [queued]>
    [2022-11-06, 08:09:27 +07] {taskinstance.py:1171} INFO - Dependencies all met for <TaskInstance: training_pipeline.model_training_task scheduled__2022-01-01T00:00:00+00:00 [queued]>
    [2022-11-06, 08:09:27 +07] {taskinstance.py:1368} INFO - 
    --------------------------------------------------------------------------------
    [2022-11-06, 08:09:27 +07] {taskinstance.py:1369} INFO - Starting attempt 1 of 1
    [2022-11-06, 08:09:27 +07] {taskinstance.py:1370} INFO - 
    --------------------------------------------------------------------------------
    [2022-11-06, 08:09:27 +07] {taskinstance.py:1389} INFO - Executing <Task(DockerOperator): model_training_task> on 2022-01-01 00:00:00+00:00
    [2022-11-06, 08:09:27 +07] {standard_task_runner.py:52} INFO - Started process 859 to run task
    [2022-11-06, 08:09:27 +07] {standard_task_runner.py:79} INFO - Running: ['***', 'tasks', 'run', 'training_pipeline', 'model_training_task', 'scheduled__2022-01-01T00:00:00+00:00', '--job-id', '9', '--raw', '--subdir', 'DAGS_FOLDER/training_pipeline/training_dag.py', '--cfg-path', '/tmp/tmpdb5dbgdq', '--error-file', '/tmp/tmpihq98uki']
    [2022-11-06, 08:09:27 +07] {standard_task_runner.py:80} INFO - Job 9: Subtask model_training_task
    [2022-11-06, 08:09:27 +07] {task_command.py:371} INFO - Running <TaskInstance: training_pipeline.model_training_task scheduled__2022-01-01T00:00:00+00:00 [running]> on host f911604fac42
    [2022-11-06, 08:09:27 +07] {taskinstance.py:1583} INFO - Exporting the following env vars:
    AIRFLOW_CTX_DAG_OWNER=mlopsvn
    AIRFLOW_CTX_DAG_ID=training_pipeline
    AIRFLOW_CTX_TASK_ID=model_training_task
    AIRFLOW_CTX_EXECUTION_DATE=2022-01-01T00:00:00+00:00
    AIRFLOW_CTX_TRY_NUMBER=1
    AIRFLOW_CTX_DAG_RUN_ID=scheduled__2022-01-01T00:00:00+00:00
    [2022-11-06, 08:09:28 +07] {docker.py:262} INFO - Starting docker container from image mlopsvn/mlops_crash_course/training_pipeline:latest
    [2022-11-06, 08:09:28 +07] {docker.py:273} WARNING - Using remote engine or docker-in-docker and mounting temporary volume from host is not supported. Falling back to `mount_tmp_dir=False` mode. You can set `mount_tmp_dir` parameter to False to disable mounting and remove the warning
    [2022-11-06, 08:09:30 +07] {docker.py:324} INFO - 2022-11-06, 08:09:30 +07 - model_training - INFO - start train_model
    2022-11-06, 08:09:30 +07 - model_training - INFO - current dir: /training_pipeline/src
    2022-11-06, 08:09:30 +07 - model_training - INFO - inspect_dir /training_pipeline/src
    2022-11-06, 08:09:30 +07 - model_training - INFO - __pycache__
    2022-11-06, 08:09:30 +07 - model_training - INFO - data_extraction.py
    2022-11-06, 08:09:30 +07 - model_training - INFO - data_preparation.py
    2022-11-06, 08:09:30 +07 - model_training - INFO - data_validation.py
    2022-11-06, 08:09:30 +07 - model_training - INFO - model_evaluation.py
    2022-11-06, 08:09:30 +07 - model_training - INFO - model_training.py
    2022-11-06, 08:09:30 +07 - model_training - INFO - model_validation.py
    2022-11-06, 08:09:30 +07 - model_training - INFO - utils.py
    2022-11-06, 08:09:30 +07 - model_training - INFO - config: {'random_seed': 17, 'feature_dict': {'conv_rate': <class 'numpy.float64'>, 'acc_rate': <class 'numpy.float64'>, 'avg_daily_trips': <class 'numpy.int64'>, 'trip_completed': <class 'numpy.int64'>}, 'target_col': 'trip_completed', 'test_size': 0.2, 'experiment_name': 'elastic-net', 'mlflow_tracking_uri': 'http://host.docker.internal:5001', 'alpha': 0.5, 'l1_ratio': 0.1, 'rmse_threshold': 0.2, 'mae_threshold': 0.2, 'registered_model_name': 'sklearn-elastic-net-reg'}
    [2022-11-06, 08:09:31 +07] {docker.py:324} INFO - 2022/11/06 01:09:31 WARNING mlflow.utils.git_utils: Failed to import Git (the Git executable is probably not on your PATH), so Git SHA is not available. Error: Failed to initialize: Bad git executable.
    The git executable must be specified in one of the following ways:
        - be included in your $PATH
        - be set via $GIT_PYTHON_GIT_EXECUTABLE
        - explicitly set via git.refresh()
    
    All git commands will error until this is rectified.
    
    This initial warning can be silenced or aggravated in the future by setting the
    $GIT_PYTHON_REFRESH environment variable. Use one of the following values:
        - quiet|q|silence|s|none|n|0: for no warning or exception
        - warn|w|warning|1: for a printed warning
        - error|e|raise|r|2: for a raised exception
    
    Example:
        export GIT_PYTHON_REFRESH=quiet
    [2022-11-06, 08:09:31 +07] {docker.py:324} INFO - 2022-11-06, 08:09:31 +07 - model_training - INFO - ('http://host.docker.internal:5001', './artifacts/4/c958513e6edf47f6b7c4329bbcdfcb46/artifacts')
    [2022-11-06, 08:09:31 +07] {docker.py:324} INFO - 2022-11-06, 08:09:31 +07 - model_training - INFO - start load_df /training_pipeline/artifacts/train_x.parquet
    [2022-11-06, 08:09:31 +07] {docker.py:324} INFO - 2022-11-06, 08:09:31 +07 - model_training - INFO - start load_df /training_pipeline/artifacts/train_y.parquet
    [2022-11-06, 08:09:31 +07] {docker.py:324} INFO - 2022/11/06 01:09:31 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/opt/venv/lib/python3.8/site-packages/mlflow/models/signature.py:129: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details."
    [2022-11-06, 08:09:35 +07] {docker.py:324} INFO - 2022-11-06, 08:09:35 +07 - model_training - INFO - Logged data and model in run c958513e6edf47f6b7c4329bbcdfcb46
    [2022-11-06, 08:09:35 +07] {docker.py:324} INFO - 2022-11-06, 08:09:35 +07 - model_training - INFO - 
    ---------- logged params ----------
    2022-11-06, 08:09:35 +07 - model_training - INFO - {'alpha': '0.5', 'copy_X': 'True', 'fit_intercept': 'True', 'l1_ratio': '0.1', 'max_iter': '1000', 'normalize': 'deprecated', 'positive': 'False', 'precompute': 'False', 'random_state': '17', 'selection': 'cyclic', 'tol': '0.0001', 'warm_start': 'False'}
    [2022-11-06, 08:09:35 +07] {docker.py:324} INFO - 2022-11-06, 08:09:35 +07 - model_training - INFO - 
    ---------- logged metrics ----------
    [2022-11-06, 08:09:35 +07] {docker.py:324} INFO - 2022-11-06, 08:09:35 +07 - model_training - INFO - {'training_mse': 0.020391301867212934, 'training_mae': 0.1339089111891788, 'training_r2_score': 0.9184347925311482, 'training_rmse': 0.14279811576912677, 'training_score': 0.9184347925311482}
    2022-11-06, 08:09:35 +07 - model_training - INFO - 
    ---------- logged tags ----------
    [2022-11-06, 08:09:35 +07] {docker.py:324} INFO - 2022-11-06, 08:09:35 +07 - model_training - INFO - {'estimator_name': 'ElasticNet', 'estimator_class': 'sklearn.linear_model._coordinate_descent.ElasticNet'}
    [2022-11-06, 08:09:35 +07] {docker.py:324} INFO - 2022-11-06, 08:09:35 +07 - model_training - INFO - 
    ---------- logged artifacts ----------
    2022-11-06, 08:09:35 +07 - model_training - INFO - ['model/MLmodel', 'model/conda.yaml', 'model/model.pkl', 'model/python_env.yaml', 'model/requirements.txt']
    [2022-11-06, 08:09:35 +07] {docker.py:324} INFO - 2022-11-06, 08:09:35 +07 - model_training - INFO - inspect_dir /training_pipeline/artifacts/run_info.json
    [2022-11-06, 08:09:35 +07] {docker.py:324} INFO - 2022-11-06, 08:09:35 +07 - model_training - INFO - Path /training_pipeline/artifacts/run_info.json is file
    [2022-11-06, 08:09:35 +07] {docker.py:324} INFO - /opt/venv/lib/python3.8/site-packages/mlflow/models/signature.py:129: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
      inputs = _infer_schema(model_input)
    [2022-11-06, 08:09:35 +07] {taskinstance.py:1412} INFO - Marking task as SUCCESS. dag_id=training_pipeline, task_id=model_training_task, execution_date=20220101T000000, start_date=20221106T010927, end_date=20221106T010935
    [2022-11-06, 08:09:35 +07] {local_task_job.py:156} INFO - Task exited with return code 0
    [2022-11-06, 08:09:35 +07] {local_task_job.py:279} INFO - 0 downstream tasks scheduled from follow-on schedule check
    ```
    

- However if we do the training directly from Python code:
    
    ```bash
    export TRAINING_PIPELINE_DIR=`pwd`
    export MLFLOW_TRACKING_URI="http://localhost:5000"
    cd src
    python model_training.py
    ```
    

It will be output as expected (artifacts are saved in `mlflow-artifacts:/`):

```bash
model_training - INFO - ('http://localhost:5001', 'mlflow-artifacts:/1/7c54f778796448ccbd2e027ae11004b3/artifacts')
```

All experiments will be logged in MLFlow Tracking Server as well.

**Specs**

- MacOS: Ventura 13.0 on Macbook Pro M1 2020
- Docker Desktop: 4.13.1
- Docker Compose: v2.12.1
- Python: 3.9 & 3.10

**Comment by nguyen-dhn** [comment]
I have found a way to fix this issue.

First, we have to change the MLFlow version from 1.28.0 to 1.29.0 (in `mlflow/.env`). The motivation is from [this thread](https://discord.com/channels/@me/1038297874464784394/1038973554940727316). 

Second, remove the `--artifacts-destination ${ARTIFACTS_RELATIVE_PATH}` argument when running the MLFlow server (in `mlflow/mlflow/Dockerfile`).

Rebuild all images related to MLflow  and run again.

Now, the artifact's location when we run with Airflow DAGs or with python script is always the default (`mlflow-artifacts:/`).

**Comment by dao-duc-tung** [comment]
Tested for entire course using:
- mlflow 1.29.0
- python 3.9
